# <center> FP2–V5. Brzina svjetlosti
## <center> Matija Pajas

In [1]:
from scipy import *
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from scipy.optimize import curve_fit

#from sympy import *
#init_printing()

from IPython.display import HTML
from IPython.display import display

### Zadatak 1
Odredite brzinu svjetlosti u zraku.

Mjereni podaci mogu se pronaći u tablici niže. Iz podataka je izračunata srednja vrijednost razlike položaja i nepouzdanost pomoću standardne devijacije.

In [20]:
dx = {"$\Delta x$ / $\mathrm{cm}$":[130,137,132]}
df = pd.DataFrame(data = dx, index=range(1,4))
#dftohtml = df.to_html(col_space=12,decimal=",")
#print(dftohtml)
df

$\Delta x$ / $\mathrm{cm}$
1                         130
2                         137
3                         132

In [24]:
means = df.mean(0)
devs = df.std(0)
uncs = devs/sqrt(5)
df2 = pd.DataFrame([(str(means[i]) + " ± " + str(uncs[i])) for i in range(len(means))],index=list(means.keys()),columns=["Rezultat"])
df2

Rezultat
$\Delta x$ / $\mathrm{cm}$  133.0 ± 1.6124515496597098

$$\Delta x = (133 ± 1,6124515496597098) \ \mathrm{cm}$$

$$\Delta x = (1,33 ± 0.02) \ \mathrm{m}$$

Taj podatak koristi se za računanje brzine svjetlosti izrazom iz pripreme:

$$c=4f\Delta x$$

za koji treba izračunati i pripadnu nepouzdanost:

$$
\begin{align}
M_{c} & = \sqrt{ \left ( \frac{\partial \Delta x}{\partial x}M_x  \right )^2  } \\
M_{c} & = \sqrt{ \left( 4fM_x \right)^2 } \\
M_{c} & = 4f\Delta x\sqrt{\left( \frac{{M_x}^2}{{\Delta x}^2}  \right)} = 4008000 \ \mathrm{m/s}
\end{align}
$$

### Rezultat:

$$c = (266532000 \pm 4008000) \ \mathrm{m/s}$$
Zaokruženo:
$$c = (2,67 \pm 0,04) \times 10^{8} \ \mathrm{m/s}$$

In [4]:
Mc = 4*(50.1*10**6)*1.33*((0.02)/(1.33))
print(Mc)

4008000.0


In [5]:
c = 4*(50.1*10**6)*1.33
print("v =",c, "m/s")
print("v =",c/(3*10**8),"c")

v = 266532000.0 m/s
v = 0.88844 c


### Zadatak 2
Odredite brzinu svjetlosti u vodi ili sintetskoj smoli (sami izaberite) te izračunajte pripadni indeks loma.

Duljina bloka smole kroz koju se pušta svjetlost za ovaj zadatak izmjerena je $l = 29,5 \ \mathrm{cm}$ uz procijenjenu nepouzdanost $M_l = 0,5 \ \mathrm{cm}$ te se mjeri nova razlika položaja $\Delta x$. Izmjereni podaci mogu se naći u tablici niže, kao i izračunata srednja vrijednost i nepouzdanost.

In [15]:
dx = {"$\Delta x$ / $\mathrm{cm}$":[8,7.5,10]}

df = pd.DataFrame(data = dx, index=range(1,4))
#dftohtml = df.to_html(col_space=12,decimal=",")
#print(dftohtml)
df

$\Delta x$ / $\mathrm{cm}$
1                         8.0
2                         7.5
3                        10.0

In [18]:
means = df.mean(0)
devs = df.std(0)
uncs = devs/sqrt(2)
df2 = pd.DataFrame([(str(means[i]) + " ± " + str(uncs[i])) for i in range(len(means))],index=list(means.keys()),columns=["Rezultat"])
df2

Rezultat
$\Delta x$ / $\mathrm{cm}$  8.5 ± 0.9354143466934853

$$\Delta x = (8.5 ± 0.5916079783099616) \ \mathrm{cm}$$

$$\Delta x = (0,085 ± 0,006) \ \mathrm{cm}$$

Da bi se došlo do brzine svjetlosti u smoli prije svega potrebno je izračunati indeks loma pomoću izraza iz pripreme:

$$n_\mathrm{m}=\frac{c_\mathrm{z}}{c_\mathrm{m}}=1+\frac{2 \Delta x}{\kappa l_\mathrm{m}}$$

gdje $\kappa$ označava broj prolazaka svjetlosti kroz duljinu bloka smole. U ovome slučaju svjetlost kroz smolu prolazi samo u jednom smjeru tako da je $\kappa = 1$ i izraz se svodi na idući izraz i iznosi:

$$n_\mathrm{m}=1+\frac{2 \Delta x}{l_\mathrm{m}} = 1,576271186440678$$

a njegova pripadna nepouzdanost je:

$$
\begin{align}
M_{n_\mathrm{m}} & =\sqrt{ \left ( \frac{ \partial n_\mathrm{m}}{\partial \Delta x}M_{\Delta x}  \right )^2  +  \left (\frac{ \partial n_\mathrm{m}}{\partial l_\mathrm{m}}M_{l_\mathrm{m}}  \right )^2 } \\
M_{n_\mathrm{m}} & =  \frac{2 \Delta x}{l_m}
\sqrt{ \left ( \frac{M_{\Delta x}}{\Delta x}  \right )^2 + \left( \frac{M_{l_\mathrm{m}}}{l_\mathrm{m}} \right )^2} = 0,04183416350925095
\end{align}
$$

### Rezultat:

$$n_\mathrm{m} = (1,576271186440678 \pm 0,04183416350925095)$$

$$n_\mathrm{m} = (1,58 \pm 0,04)$$

Sada kada je indeks loma izračunat, može se doći do brzine idućim izrazom:

$$v_{\mathrm{s}}=\frac{c}{n_{\mathrm{s}}}= 189873417,721519 \ \mathrm{m/s}$$

a nepouzdanost brzine će biti:

$$
\begin{align}
M_{v_{\mathrm{s}}} & = \sqrt{ \left ( \frac{\partial v_\mathrm{s}}{\partial c}M_c    \right )^2 +
\left ( \frac{\partial v_\mathrm{s}}{\partial n_{\mathrm{s}}}M_{n_{\mathrm{s}}}    \right )^2} \\
M_{v_{\mathrm{s}}} & = v_\mathrm{s} \cdot
\sqrt{\left ( \frac{M_{c}}{c}  \right )^2 +
\left ( \frac{M_{n_\mathrm{s}}}{n_\mathrm{s}}  \right )^2} = 5585511,469997809 \ \mathrm{m/s}
\end{align}
$$

### Rezultat:

$$v_\mathrm{s} = (189873417,721519 \pm 005585511,469997809) \ \mathrm{m/s}$$

$$v_\mathrm{s} = (1,9 \pm 0,06) \times 10^{8} \ \mathrm{m/s}$$

In [8]:
n = 1 + (2 * 8.5)/29.5
print("n =",n)
v = (3*10**8)/n
print("v =",v,"m/s")
print("v =",v/(3*10**8),"c")
print((n-1)*sqrt((0.006**2)/(0.085**2) + (0.005**2)/(0.295**2)))
print((3*10**8)/1.58)
print(189873417.721519 * sqrt( ((0.04)/(2.67))**2 + ((0.04)/(1.58))**2 ) )

n = 1.576271186440678
v = 190322580.6451613 m/s
v = 0.6344086021505376 c
0.04183416350925095
189873417.721519
5585511.469997809


### Zadatak 3
Odredite indeks loma staklene prizme za spektralne linije iz emisijskog spektra dostupne lampe te
izračunajte pripadne brzine svjetlosti.

In [9]:
b = 5.25
v = 4.54
phi = 2*arctan((b/2)/(v))
print("phi =",phi,"rad")
degrees(phi)
print("phi =",degrees(phi),"°")

phi = 1.0484624341137057 rad
phi = 60.07247245272848 °


In [10]:
dc = radians(45.4)
dz = radians(47.5)
dp = radians(49.1)
nc = sin((phi + dc)/2)/sin(phi/2)
nz = sin((phi + dz)/2)/sin(phi/2)
np = sin((phi + dp)/2)/sin(phi/2)
print(nc,nz,np)
c = 3*10**8
vc = c/nc
vz = c/nz
vp = c/np
print(vc,vz,vp)

1.5899717778952585 1.6118714684370343 1.6281939991742844
188682594.35216397 186119058.42027074 184253227.90290394
